---
title: "R Notebook"
output: html_notebook
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

if(!require('pacman')) {install.packages('pacman')}
pacman::p_load(openxlsx, dplyr, reshape2, stringr)

# BiocManager::install("GenomicRanges")
library(GenomicRanges)
```

```{r}
source('LibraryDesignFunctions.R')
```


# Input Parameters + Load Data
```{r}
# Import Relevant Datasets
wd <- getwd()
wd # verify directory location

snp_list <- unique(read.xlsx(paste0(wd,"//input//snp_loc_hg19.xlsx")))
rownames(snp_list) <- NULL
snp_list$chr <- as.character(snp_list$chr)


# Get refseq conversion for chromosome numbers. Needed for compatibility with CRISPick
refseqLookup <- read.delim(paste0(wd,"//input//seq_report_hg19.tsv"))
refseqLookup <- subset(refseqLookup, Role=="assembled-molecule")

# Rename columns for convenience
names(refseqLookup)[names(refseqLookup)=="Chromosome.name"] <- "chr"
names(refseqLookup)[names(refseqLookup)=="RefSeq.seq.accession"] <- "refseq"
refseqLookup <- refseqLookup %>%
  select(chr, refseq)
```

# Preprocess Input for CRISPick

## Combine Nearby SNP Regions

Will use 200bp regions for most SNPs. The 300 and 400bp regions are for any
200bp regions that don't have enough good quality guides.
```{r}
regions200 <- findSnpRegions(200, snp_list)
regions300 <- findSnpRegions(300, snp_list)
regions400 <- findSnpRegions(400, snp_list)

# any(is.na(g_overlap@to)) # << Check for lost SNPs. Should be FALSE
```
## Convert to CRISPick Input Format
```{r}
crispick200 <- formatCrispick(regions200, refseqLookup)
crispick300 <- formatCrispick(regions300, refseqLookup)
crispick400 <- formatCrispick(regions400, refseqLookup)

# Only need the 'crispick' column to run CRISPick for library design.
crispick200Input <- crispick200 %>% select(crispick)
crispick300Input <- crispick300 %>% select(crispick)
crispick400Input <- crispick400 %>% select(crispick)
```

## Export SNP Regions as regions and as CRISPick Input Format
```{r}
# Save SNP regions
write.xlsx(regions200, file = paste0(wd, "//preprocessed//regions200.xlsx"))
write.xlsx(regions300, file = paste0(wd, "//preprocessed//regions300.xlsx"))
write.xlsx(regions400, file = paste0(wd, "//preprocessed//regions400.xlsx"))

# Save CRISPick Inputs
write.table(crispick200Input, paste0(wd, "//preprocessed//loc200.fasta"), row.names=FALSE, col.names=FALSE, quote=FALSE)
write.table(crispick300Input, paste0(wd, "//preprocessed//loc300.fasta"), row.names=FALSE, col.names=FALSE, quote=FALSE)
write.table(crispick400Input, paste0(wd, "//preprocessed//loc400.fasta"), row.names=FALSE, col.names=FALSE, quote=FALSE)
```

PAUSE HERE >> run CRISPick using fasta file.

----------------------------------------------------------------------------

# Create Library from CRISPick Output
## Import CRISPick Output
```{r}
GDO200 <- read.xlsx(paste0(wd,"//crispick_output//sgrna_designs_200bp_i30.xlsx"))
GDO300 <- read.xlsx(paste0(wd,"//crispick_output//sgrna_designs_300bp_i30.xlsx"))
GDO400 <- read.xlsx(paste0(wd,"//crispick_output//sgrna_designs_400bp_i30.xlsx"))

keepCol <- c("Input", "Reference.Sequence", "Orientation", "sgRNA.'Cut'.Position", 
            "sgRNA.Sequence", "sgRNA.Context.Sequence", "On-Target.Efficacy.Score", "Pick.Order")

GDO200 <- select(GDO200, all_of(keepCol))
GDO300 <- select(GDO300, all_of(keepCol))
GDO400 <- select(GDO400, all_of(keepCol))

# Alternative column selection:
# guides <- guides %>%
#   subset(select = -c(Quota, Target.Taxon, Target.Gene.ID, Target.Gene.Symbol, Target.Alias, 
#                      CRISPR.Mechanism, Target.Domain, TSS.Position, PAM.Policy))

```

## GC + TTTT filter
Good GC content, no poly T, On-Target Efficacy > 0.2
Renumbers the rows after removing guides that don't meet the filter criteria.
```{r}
GDO200 <- filterGDO(GDO200)
GDO300 <- filterGDO(GDO300)
GDO400 <- filterGDO(GDO400)
```

## Add guide coordinates
```{r}
# Add GDO locations: chromosome number + start/end coordinates.
GDO200 <- addGDOLoc(GDO200, refseqLookup)
GDO300 <- addGDOLoc(GDO300, refseqLookup)
GDO400 <- addGDOLoc(GDO400, refseqLookup)

# Add the SNP Region
GDO200 <- left_join(GDO200, select(crispick200, SNP, crispick), join_by("Input"=="crispick"))
GDO300 <- left_join(GDO300, select(crispick300, SNP, crispick), join_by("Input"=="crispick"))
GDO400 <- left_join(GDO400, select(crispick400, SNP, crispick), join_by("Input"=="crispick"))

# Add rownumber for reference later
GDO200$n <- as.numeric(rownames(GDO200))
GDO300$n <- as.numeric(rownames(GDO300))
GDO400$n <- as.numeric(rownames(GDO400))

# Debug
# all(GDO200$start < GDO200$end) # CHECK that coordinate order is correct. TRUE if okay.
# all(guides$n == as.numeric(rownames(guides))) #CHECK that n == rownumber. TRUE if okay.

```

## Get top non-overlapping guides.

Get indices of overlapping guides.
```{r}
GDO200Overlap <- findGDOSelfOverlap(GDO200, "chr", "start", "end")
GDO300Overlap <- findGDOSelfOverlap(GDO300, "chr", "start", "end")
GDO400Overlap <- findGDOSelfOverlap(GDO400, "chr", "start", "end")
```

Get top guides.
```{r}
# Pick the top performing guides for each SNP region width.
# Here we pick the top 5. Could pick top 3 etc.
numGDO <- 3

topGDO200 <- pickTopGDO(numGDO, GDO200, regions200, GDO200Overlap)
topGDO300 <- pickTopGDO(numGDO, GDO300, regions300, GDO300Overlap)
topGDO400 <- pickTopGDO(numGDO, GDO400, regions400, GDO400Overlap)
```

Export list of top guides.
```{r}
write.xlsx(topGDO200, file = paste0(wd, "//output//top3_200bp_i30.xlsx"))
write.xlsx(topGDO300, file = paste0(wd, "//output//top3_300bp_i30.xlsx"))
write.xlsx(topGDO400, file = paste0(wd, "//output//top3_400bp_i30.xlsx"))
```


# QC
How is the on-target score?
```{r}
print("Mean Number of Guides per SNP")
print(paste("GDO200:", mean(table(topGDO200$SNP))))
print(paste("GDO300:", mean(table(topGDO300$SNP))))
print(paste("GDO400:", mean(table(topGDO400$SNP))))

print("Mean On-Target Efficacy Score for the Library")
print(paste("GDO200:", mean(topGDO200$`On-Target.Efficacy.Score`)))
print(paste("GDO300:", mean(topGDO300$`On-Target.Efficacy.Score`)))
print(paste("GDO400:", mean(topGDO400$`On-Target.Efficacy.Score`)))

hist(table(topGDO200$SNP))
hist(table(topGDO300$SNP))
hist(table(topGDO400$SNP))

hist(topGDO200$`On-Target.Efficacy.Score`)
hist(topGDO300$`On-Target.Efficacy.Score`)
hist(topGDO400$`On-Target.Efficacy.Score`)

# On-Target Efficacy by SNP Region
# guide_summary <- top_guides %>%
#   group_by(SNP) %>%
#   summarise(mean = mean(`On-Target.Efficacy.Score`))
# 
# subset(guide_summary, mean < 0.2)
```
# Combining Guide Searches
First ran guide search for 200bp, ran QC. Then ran 300bp and 400bp to get more guides for poorly performing regions.

Load all GDOs + Regions if needed:
```{r}
# # Load all guide pools
# GDO200 <- read.xlsx(paste0(wd,"//output//top3_200bp_i30.xlsx"))
# GDO300 <- read.xlsx(paste0(wd,"//output//top3_300bp_i30.xlsx"))
# GDO400 <- read.xlsx(paste0(wd,"//output//top3_400bp_i30.xlsx"))
# # note, less guides in 300bp & 400bp because the regions were combined. There are fewer regions.
# 
# # Load regions by width
# regions200 <- read.xlsx(paste0(wd,"//preprocessed//regions200.xlsx"))
# regions300 <- read.xlsx(paste0(wd,"//preprocessed//regions300.xlsx"))
# regions400 <- read.xlsx(paste0(wd,"//preprocessed//regions400.xlsx"))
```

Loading all guide searches:
```{r}
# Initialize final library
regions <- regions200
finalGDO <- GDO200
numGDO <- 3

# Get any regions that are missing or do not have enough guides
badRegion <- getBadRegion(finalGDO, regions, numGDO) # 3 bad regions

if(nrow(badRegion)>0){
  finalGDO <- replaceBadRegion(finalGDO, GDO300, regions, badRegion)
}

badRegion <- getBadRegion(finalGDO, regions, numGDO) # 2 bad regions

if(nrow(badRegion)>0){
  finalGDO <- replaceBadRegion(finalGDO, GDO400, regions, badRegion)
}

badRegion <- getBadRegion(finalGDO, regions, numGDO) # no bad regions
```

Quality Control
```{r}
# Check that you have all the expected regions from the 200bp
  # i.e. you haven't LOST any regions and you haven't ADDED any regions by accident

all(final_GDO$SNP %in% regions$SNP)
!any(is.na(final_GDO$SNP))


# Check for no overlaps between GUIDES in the library

g_GDO <- GRanges(seqnames = final_GDO$chr,
                    ranges = IRanges(start = final_GDO$start, end=final_GDO$end))

g_overlap <- findOverlaps(g_GDO, ignore.strand=TRUE, drop.self=TRUE)

GDO_overlap <- as.data.frame(g_overlap) # no overlaps detected

# Check that each guide overlaps with only ONE REGION

regions <- regions_200

g_regions <- GRanges(seqnames = regions$chr, 
                     ranges = IRanges(start = regions$start, end=regions$end))

g_overlap <- findOverlaps(g_GDO, g_regions, ignore.strand=TRUE)

region_overlap <- as.data.frame(g_overlap) # 212 out of 222 guides overlap the 200bp region
```


Export guide list:
```{r}
# write.xlsx(final_GDO, file = paste0(wd, "//output//final_GDO_v1.xlsx"))

GDO_output <- read.xlsx(paste0(wd,"//output//final_GDO_v1.xlsx"))

# all(GDO_output$sgRNA.Sequence %in% final_GDO$sgRNA.Sequence)
# all(final_GDO$sgRNA.Sequence %in% GDO_output$sgRNA.Sequence)
```

# Guide Library Updates

Add + Replace Positive Controls

## Load data
```{r}
wd <- getwd()

# Import new guide designs
posA30 <- read.xlsx(paste0(wd,"//crispick_output//pos_ctrl_a30_sgrna-designs.xlsx"))
names(posA30)[names(posA30)=="Target.Gene.Symbol"] <- "SNP"


```

```{r}
# Add chromosome number based on refseq lookup table
lookup <- refseq_lookup %>%
  select(chr, refseq)

posA30 <- left_join(posA30, lookup, join_by("Reference.Sequence"=="refseq"))

# Placeholder columns for start and end..
posA30$start <- NA
posA30$end <- NA


# probably could do mutate() for this..
for (i in 1:nrow(posA30)){
  # if the guide is sense vs antisense
  if (posA30$Orientation[i]=="sense"){
    posA30$start[i] = posA30$`sgRNA.'Cut'.Position`[i] - 17
    posA30$end[i] = posA30$`sgRNA.'Cut'.Position`[i] + 2
    
  } else if (posA30$Orientation[i]=="antisense"){
    posA30$start[i] = posA30$`sgRNA.'Cut'.Position`[i] - 3
    posA30$end[i] = posA30$`sgRNA.'Cut'.Position`[i] + 16
    
  }
}

# Use Input as the "SNP" region
# posA30 <- left_join(posA30, select(crispick_reg, SNP, crispick), join_by("Input"=="crispick"))
names(posA30)[names(posA30)=="Target.Gene.Symbol"] <- "SNP"

all(posA30$start < posA30$end) # CHECK that coordinate order is correct. TRUE if okay.

# Add rownumber for reference later
posA30$n <- as.numeric(rownames(posA30))


all(posA30$n == as.numeric(rownames(posA30))) #CHECK that n == rownumber. TRUE if okay.
```


```{r}
posGene <- read.table(paste0(wd,"//input//crispick_pos_ctrl.txt"))
colnames(posGene) <- "SNP"

validCtrl <- read.xlsx(paste0(wd, "//input//validated_pos_ctrl.xlsx"))
```

Will use guidees from pos_a30 only.
First remove any guides that overlap the validated guides.

```{r}
gValid <- GRanges(seqnames = validCtrl$chr, 
                     ranges = IRanges(start = validCtrl$start, end=validCtrl$end))

gCrisp <- GRanges(seqnames = posA30$chr, 
                     ranges = IRanges(start = posA30$start, end=posA30$end))

gOverlap <- findOverlaps(gValid, gCrisp, ignore.strand=TRUE)

region_overlap <- as.data.frame(gOverlap) # ~ 12 guides overlap the validated guides

```

Remove overlapping guides
```{r}
# remove <- region_overlap[region_overlap$queryHits==top$n,2] 
posPool <- posA30[!posA30$n %in% region_overlap[,2],]
```




```{r}

gCrisp <- GRanges(seqnames = posA30$chr, 
                     ranges = IRanges(start = posA30$start, end=posA30$end)) # forgot to add chr, start and stop

gOverlap <- findOverlaps(gCrisp, ignore.strand=TRUE)

region_overlap <- as.data.frame(gOverlap) # 212 out of 222 guides overlap the 200bp region
# none of the new CRISPick guides overlap with the validated guides.

topPos <- pickTopGDO(3, posA30, posGene, region_overlap)
```

Export the top guides. Go in order of pick order, fill in where we don't have validated guides.
Will combine the lists manually since there's not too many guides.
Need to manually pick+evaluate RPA1/SMYD4 guides since the region is overlapping..
```{r}
write.xlsx(topPos, file = paste0(wd, "//output//pos_ctrl_v1.xlsx"))
```


# Misc
Comparing CRISPRi vs CRISPRa for CRISPick
```{r}
# top_guides_i <- top_guides
# top_guides_a <- top_guides

top_diff <- anti_join(top_guides_i, top_guides_a, by="sgRNA.Sequence")
```

