---
title: "R Notebook"
output: html_notebook
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

if(!require('pacman')) {install.packages('pacman')}
pacman::p_load(openxlsx, dplyr, reshape2, stringr)

# BiocManager::install("GenomicRanges")
library(GenomicRanges)
```

```{r}
source('LibraryDesignFunctions.R')
```


# Input Parameters + Load Data
```{r}
# Import Relevant Datasets
wd <- getwd()
wd # verify directory location

snp_list <- unique(read.xlsx(paste0(wd,"//input//snp_loc_hg19.xlsx")))
rownames(snp_list) <- NULL
snp_list$chr <- as.character(snp_list$chr)

# Get refseq conversion for chromosome numbers. Needed for compatibility with CRISPick
refseqLookup <- read.delim(paste0(wd,"//input//seq_report_hg19.tsv"))
refseqLookup <- subset(refseq_lookup, Role=="assembled-molecule")

# Rename columns for convenience
names(refseqLookup)[names(refseqLookup)=="Chromosome.name"] <- "chr"
names(refseqLookup)[names(refseqLookup)=="RefSeq.seq.accession"] <- "refseq"
```
# Preprocess Input for CRISPick
## Combine Nearby SNP Regions
```{r}
regions200 <- findSnpRegions(200, snp_list)
regions300 <- findSnpRegions(300, snp_list)
regions400 <- findSnpRegions(400, snp_list)

# any(is.na(g_overlap@to)) # << Check for lost SNPs. Should be FALSE
```
## Convert to CRISPick Input Format
```{r}
crispick200 <- formatCrispick(regions200, refseqLookup)
crispick300 <- formatCrispick(regions300, refseqLookup)
crispick400 <- formatCrispick(regions400, refseqLookup)

crispick200Input <- crispick200 %>% select(crispick)
crispick300Input <- crispick300 %>% select(crispick)
crispick400Input <- crispick400 %>% select(crispick)
```

## Export SNP Regions as regions and as CRISPick Input Format
```{r}
# Save SNP regions
write.xlsx(regions, file = paste0(wd, "//input//regions_400bp.xlsx"))
write.xlsx(regions, file = paste0(wd, "//input//regions_400bp.xlsx"))
write.xlsx(regions, file = paste0(wd, "//input//regions_400bp.xlsx"))

# Save CRISPick Inputs
write.table(crispick200Input, "loc200.fasta", row.names=FALSE, col.names=FALSE, quote=FALSE)
write.table(crispick300Input, "loc200.fasta", row.names=FALSE, col.names=FALSE, quote=FALSE)
write.table(crispick400Input, "loc200.fasta", row.names=FALSE, col.names=FALSE, quote=FALSE)
```

PAUSE HERE >> run CRISPick using fasta file.

----------------------------------------------------------------------------

# Create Library from CRISPick Output
## Import CRISPick Output
```{r}
guides <- read.xlsx(paste0(wd,"//200bp_i30//sgrna-designs.xlsx"))

#Remove some extraneous columns
# guides <- guides %>%
#   subset(select = -c(Quota, Target.Taxon, Target.Gene.ID, Target.Gene.Symbol, Target.Alias, 
#                      CRISPR.Mechanism, Target.Domain, TSS.Position, PAM.Policy))

# For debugging
guides <- guides %>%
  select( c(Input, Reference.Sequence, Orientation, `sgRNA.'Cut'.Position`, 
            sgRNA.Sequence, sgRNA.Context.Sequence, `On-Target.Efficacy.Score`, Pick.Order) )

```

## GC + TTTT filter
Good GC content, no poly T, On-Target Efficacy > 0.2
```{r}
guides <- guides %>%
  mutate(GC_percent = (str_count(guides$sgRNA.Sequence,"G")+str_count(guides$sgRNA.Sequence,"C"))/20,
         polyT = str_count(guides$sgRNA.Sequence,"TTTT")) # TTT shows up, so this seems to work.

guides <- guides %>%
  subset((GC_percent>=0.25) & (GC_percent<=0.75) & (polyT < 1) & (`On-Target.Efficacy.Score` > 0.2)) #filtered ~40 guides of 2000; for 200bp_i30 w/ efficacy filter, went from ~1.7k to 700 guides.

rownames(guides) = NULL
```

## Add guide coordinates
```{r}
# Add chromosome number based on refseq lookup table
lookup <- refseq_lookup %>%
  select(chr, refseq)

guides <- left_join(guides, lookup, join_by("Reference.Sequence"=="refseq"))

# Placeholder columns for start and end..
guides$start <- NA
guides$end <- NA


# probably could do mutate() for this..
for (i in 1:nrow(guides)){
  # if the guide is sense vs antisense
  if (guides$Orientation[i]=="sense"){
    guides$start[i] = guides$`sgRNA.'Cut'.Position`[i] - 17
    guides$end[i] = guides$`sgRNA.'Cut'.Position`[i] + 2
    
  } else if (guides$Orientation[i]=="antisense"){
    guides$start[i] = guides$`sgRNA.'Cut'.Position`[i] - 3
    guides$end[i] = guides$`sgRNA.'Cut'.Position`[i] + 16
    
  }
}

# Add the SNP Region
guides <- left_join(guides, select(crispick_reg, SNP, crispick), join_by("Input"=="crispick"))

all(guides$start < guides$end) # CHECK that coordinate order is correct. TRUE if okay.

# Add rownumber for reference later
guides$n <- as.numeric(rownames(guides))


all(guides$n == as.numeric(rownames(guides))) #CHECK that n == rownumber. TRUE if okay.

```

## Remove Overlapping Guides
Get indices of overlapping guides.
```{r}
g_guides <- GRanges(seqnames = guides$chr,
                    ranges = IRanges(start = guides$start, end=guides$end))

g_GDO_overlap <- findOverlaps(g_guides, ignore.strand=TRUE) #drop.self=TRUE, 

GDO_overlap <- as.data.frame(g_GDO_overlap)
```

Get top
```{r}
# Input: guides, num_top_guides, regions, GDO_overlap

topGDO <- pickTopGDO(5, guides, regions, GDO_overlap)

```

# QC
How is the on-target score?
```{r}
mean(table(top_guides$SNP))
hist(table(top_guides$SNP))

mean(top_guides$`On-Target.Efficacy.Score`)
hist(top_guides$`On-Target.Efficacy.Score`)

# guide_summary <- top_guides %>%
#   group_by(SNP) %>%
#   summarise(mean = mean(`On-Target.Efficacy.Score`))
# 
# subset(guide_summary, mean < 0.2)
```
How many/which regions don't have enough guides?
```{r}
# Check for any regions that dropped out completely
reg <- data.frame(table(top_guides$SNP))
names(reg) <- c("SNP", "Freq")
missing_reg <- anti_join(regions, reg, by="SNP")

missing <- data.frame(SNP = missing_reg$SNP, Freq = 0)

low_reg <- subset(reg, Freq < 3)

test <- dplyr::union(low_reg, missing) # 10 regions dont' have enough guides
```

# Export guides
```{r}
write.xlsx(guides, file = paste0(wd, "//output//guides_400bp_i30_0.2.xlsx"))

write.xlsx(top_guides, file = paste0(wd, "//output//top3guides_400bp_i30_0.2.xlsx"))
```

# Combining Guide Searches
First ran guide search for 200bp, ran QC. Then ran 300bp and 400bp to get more guides for poorly performing regions.

Loading all guide searches:
```{r}
# Load all guide pools
GDO_200bp <- read.xlsx(paste0(wd,"//output//top3guides_200bp_i30_0.2.xlsx"))
GDO_300bp <- read.xlsx(paste0(wd,"//output//top3guides_300bp_i30_0.2.xlsx"))
GDO_400bp <- read.xlsx(paste0(wd,"//output//top3guides_400bp_i30_0.2.xlsx"))
# note, less guides in 300bp+400bp because the regions were combined. There are fewer regions.

# Load regions by width
regions_200 <- read.xlsx(paste0(wd,"//input//regions_200bp.xlsx"))
regions_300 <- read.xlsx(paste0(wd,"//input//regions_300bp.xlsx"))
regions_400 <- read.xlsx(paste0(wd,"//input//regions_400bp.xlsx"))

# Initialize final library
regions <- regions_200
final_GDO <- GDO_200bp

# ----------------------------------------------------------------

# QC Checks: missing regions, few guides, or low score.
# Creates list of regions that need a bigger guide pool.

reg <- data.frame(table(final_GDO$SNP)) %>% # Create dataframe that counts number of guides/SNP
  setNames(c("SNP","Freq")) # Set column names.

poor_reg <- data.frame(SNP=character())

poor_reg <- bind_rows(poor_reg, 
                  anti_join(regions, reg, by = "SNP") %>%
                    select(SNP))

poor_reg <- bind_rows(poor_reg, reg %>% 
                        subset(Freq < 3) %>%
                        select(SNP))

# Removed Efficacy < 0.2 in Preprocessing

# poor_reg <- bind_rows(poor_reg,
#   group_by(final_GDO, SNP) %>%
#   summarise(mean = mean(`On-Target.Efficacy.Score`)) %>%
#   subset(mean < 0.2) %>%
#   select(SNP))

# ------------------------------------------------------------------
# Get guides from 300bp pool for bad regions

final_GDO <- subset(final_GDO, !(SNP %in% poor_reg$SNP))
final_GDO <- bind_rows(final_GDO, 
                   subset(GDO_300bp, SNP %in% poor_reg$SNP))

# --------------------------------------------------------------------

# Update list of regions that need more/better guides

reg <- data.frame(table(final_GDO$SNP)) %>% # Create dataframe that counts number of guides/SNP
  setNames(c("SNP","Freq")) # Set column names.

poor_reg <- data.frame(SNP=character())

poor_reg <- bind_rows(poor_reg, 
                  anti_join(regions, reg, by = "SNP") %>%
                    select(SNP))

poor_reg <- bind_rows(poor_reg, reg %>% 
                        subset(Freq < 3) %>%
                        select(SNP))


# --------------------------------------------------------------------
# Update bad regions w/ guides from the 400bp search. Matches by SNP region.

final_GDO <- subset(final_GDO, !(SNP %in% poor_reg$SNP))
final_GDO <- bind_rows(final_GDO, 
                   subset(GDO_400bp, SNP %in% poor_reg$SNP))


# --------------------------------------------------------------------

# QUALITY CONTROL

# Check that you have all the expected regions from the 200bp
  # i.e. you haven't LOST any regions and you haven't ADDED any regions by accident

all(final_GDO$SNP %in% regions$SNP)
!any(is.na(final_GDO$SNP))


# Check for no overlaps between GUIDES in the library

g_GDO <- GRanges(seqnames = final_GDO$chr,
                    ranges = IRanges(start = final_GDO$start, end=final_GDO$end))

g_overlap <- findOverlaps(g_GDO, ignore.strand=TRUE, drop.self=TRUE)

GDO_overlap <- as.data.frame(g_overlap) # no overlaps detected

# Check that each guide overlaps with only ONE REGION

regions <- regions_200

g_regions <- GRanges(seqnames = regions$chr, 
                     ranges = IRanges(start = regions$start, end=regions$end))

g_overlap <- findOverlaps(g_GDO, g_regions, ignore.strand=TRUE)

region_overlap <- as.data.frame(g_overlap) # 212 out of 222 guides overlap the 200bp region


```

Export guide list:
```{r}
# write.xlsx(final_GDO, file = paste0(wd, "//output//final_GDO_v1.xlsx"))

GDO_output <- read.xlsx(paste0(wd,"//output//final_GDO_v1.xlsx"))

# all(GDO_output$sgRNA.Sequence %in% final_GDO$sgRNA.Sequence)
# all(final_GDO$sgRNA.Sequence %in% GDO_output$sgRNA.Sequence)
```

# Guide Library Updates

Add + Replace Positive Controls

## Load data
```{r}
wd <- getwd()

# Import new guide designs
posA30 <- read.xlsx(paste0(wd,"//crispick_output//pos_ctrl_a30_sgrna-designs.xlsx"))
names(posA30)[names(posA30)=="Target.Gene.Symbol"] <- "SNP"


```

```{r}
# Add chromosome number based on refseq lookup table
lookup <- refseq_lookup %>%
  select(chr, refseq)

posA30 <- left_join(posA30, lookup, join_by("Reference.Sequence"=="refseq"))

# Placeholder columns for start and end..
posA30$start <- NA
posA30$end <- NA


# probably could do mutate() for this..
for (i in 1:nrow(posA30)){
  # if the guide is sense vs antisense
  if (posA30$Orientation[i]=="sense"){
    posA30$start[i] = posA30$`sgRNA.'Cut'.Position`[i] - 17
    posA30$end[i] = posA30$`sgRNA.'Cut'.Position`[i] + 2
    
  } else if (posA30$Orientation[i]=="antisense"){
    posA30$start[i] = posA30$`sgRNA.'Cut'.Position`[i] - 3
    posA30$end[i] = posA30$`sgRNA.'Cut'.Position`[i] + 16
    
  }
}

# Use Input as the "SNP" region
# posA30 <- left_join(posA30, select(crispick_reg, SNP, crispick), join_by("Input"=="crispick"))
names(posA30)[names(posA30)=="Target.Gene.Symbol"] <- "SNP"

all(posA30$start < posA30$end) # CHECK that coordinate order is correct. TRUE if okay.

# Add rownumber for reference later
posA30$n <- as.numeric(rownames(posA30))


all(posA30$n == as.numeric(rownames(posA30))) #CHECK that n == rownumber. TRUE if okay.
```


```{r}
posGene <- read.table(paste0(wd,"//input//crispick_pos_ctrl.txt"))
colnames(posGene) <- "SNP"

validCtrl <- read.xlsx(paste0(wd, "//input//validated_pos_ctrl.xlsx"))
```

Will use guidees from pos_a30 only.
First remove any guides that overlap the validated guides.

```{r}
gValid <- GRanges(seqnames = validCtrl$chr, 
                     ranges = IRanges(start = validCtrl$start, end=validCtrl$end))

gCrisp <- GRanges(seqnames = posA30$chr, 
                     ranges = IRanges(start = posA30$start, end=posA30$end))

gOverlap <- findOverlaps(gValid, gCrisp, ignore.strand=TRUE)

region_overlap <- as.data.frame(gOverlap) # ~ 12 guides overlap the validated guides

```

Remove overlapping guides
```{r}
# remove <- region_overlap[region_overlap$queryHits==top$n,2] 
posPool <- posA30[!posA30$n %in% region_overlap[,2],]
```




```{r}

gCrisp <- GRanges(seqnames = posA30$chr, 
                     ranges = IRanges(start = posA30$start, end=posA30$end)) # forgot to add chr, start and stop

gOverlap <- findOverlaps(gCrisp, ignore.strand=TRUE)

region_overlap <- as.data.frame(gOverlap) # 212 out of 222 guides overlap the 200bp region
# none of the new CRISPick guides overlap with the validated guides.

topPos <- pickTopGDO(3, posA30, posGene, region_overlap)
```

Export the top guides. Go in order of pick order, fill in where we don't have validated guides.
Will combine the lists manually since there's not too many guides.
Need to manually pick+evaluate RPA1/SMYD4 guides since the region is overlapping..
```{r}
write.xlsx(topPos, file = paste0(wd, "//output//pos_ctrl_v1.xlsx"))
```


# Misc
Comparing CRISPRi vs CRISPRa for CRISPick
```{r}
# top_guides_i <- top_guides
# top_guides_a <- top_guides

top_diff <- anti_join(top_guides_i, top_guides_a, by="sgRNA.Sequence")
```

