---
title: "Final Library Design"
author: Shannon Laub
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

if(!require('pacman')) {install.packages('pacman')}
pacman::p_load(openxlsx, dplyr, tidyr, reshape2, stringr)

# BiocManager::install("GenomicRanges")
library(GenomicRanges)

source('libraryDesignFunctions.R')
```

Before running this script, go through the following to generate guide pools:
- preprocessCRISPick.Rmd - to create raw guide pools
- preprocessGDO.Rmd - to preprocess the raw guide pools for this script

# Set Directory Information
```{r}
wd <- getwd()
wd # verify directory location

# Set universal directories:
dirRegion <- paste0(wd,"//preprocessed//")
dirGDO <- paste0(wd, "//preprocessed//")

snpList <- unique(read.xlsx(paste0(wd,"//input//snp_loc_hg19.xlsx")))
rownames(snpList) <- NULL
snpList$chr <- as.character(snpList$chr)
```

# Assemble Library
```{r}
regionList <- c("regions200", "regions300", "regions400", "regions500", "regions600")
fileList <- c("200bp_i30","300bp_i30","400bp_i30","500bp_i30","600bp_i30")
targetGDO <- assembleLibrary(3, fileList, regionList, dirGDO, dirRegion,
                          onTargetFilter = 0.2, pickByStrand = TRUE, overlapLeniency=2)

uniqueGDO <- unique(targetGDO)
nrow(uniqueGDO) == nrow(targetGDO)

overlap <- findSelfOverlap(targetGDO, "chr", "start","end", selfBool=TRUE)
overlapLenient <- findSelfOverlap(targetGDO, "chr", "startL","endL", selfBool=TRUE)
```

```{r}
qcSimple(targetGDO)
qcSumOff(targetGDO)
orient <- qcStrand(targetGDO)
```


Quality Control
```{r}
# Load regions for reference

regions200 <- read.xlsx(paste0(wd,"//preprocessed//regions200.xlsx"))
regions500 <- read.xlsx(paste0(wd,"//preprocessed//regions500.xlsx"))

# Check that you have all the expected regions from the 200bp
  # i.e. you haven't LOST any regions and you haven't ADDED any regions by accident

all(targetGDO$SNP %in% regions200$SNP)
!any(is.na(targetGDO$SNP))

# Check that each guide overlaps with only ONE REGION

reg200Overlap <- findOverlap(targetGDO, regions200,
                          "chr", "start", "end",
                          "chr", "start", "end")
# 201 out of 222 guides overlap the 200bp region

reg500Overlap <- findOverlap(targetGDO, regions500,
                          "chr", "start", "end",
                          "chr", "start", "end")
# All guides overlap a 500bp region.
```

Export guide list:
```{r}
write.xlsx(targetGDO, file = paste0(wd, "//output//finalTargetGDOi30.xlsx"))

# targetGDO <- read.xlsx(paste0(wd,"//output//finalTargetGDOi30.xlsx"))
```

# Substituting in Working Guides from the Pilot

Add/drop of GDOs is based on finalTargetGDOi30.xlsx.

Load list of target GDOs
```{r}
wd <- getwd()
wd # verify directory location

targetGDO <- read.xlsx(paste0(wd,"//output//finalTargetGDOi30.xlsx"))
```


Regions for rs1870138|rs7080009 and rs1870137 are ~50bp away from each other. Let's combine them.
We can reuse the guides from the Pilot Screen.

```{r}
# TSPAN14 3x SNP region:
# 1) remove sequences:
# AAGGAGGATTGCAGCTGGTA, ACAAGAGCCTTGCATTGCAC, TCCCTGGAATATGTGCCCAG
removeGDO <- c("AAGGAGGATTGCAGCTGGTA", "ACAAGAGCCTTGCATTGCAC", "TCCCTGGAATATGTGCCCAG")
targetGDO <- subset(targetGDO, !(sgRNA.Sequence %in% removeGDO))

# 2) Change region name to rs1870137|rs1870138|rs7080009 for seqs:
# CTTAGGCGCTGCATACCGTA, GGAACAAGCCACTCCACCTC, GAACTGCTTAGACAAACCCC
keepGDO <- c("CTTAGGCGCTGCATACCGTA", "GGAACAAGCCACTCCACCTC", "GAACTGCTTAGACAAACCCC")
targetGDO <- targetGDO %>%
  mutate(SNP = ifelse(sgRNA.Sequence %in% keepGDO, "rs1870137|rs1870138|rs7080009", SNP))
```

For rs9633740, the best performing guide from the pilot screen is already included.

For rs8019279,rs8019585,rs8003334 we have already tested all three guides in the new library. Let's remove the worst performing guide.

```{r}
# Remove the bad performing guide
targetGDO <- subset(targetGDO, sgRNA.Sequence != "AGAACAATAAAAACTACGGT")

# Load Guide Pool, subset for region of interest, remove guide that you don't want from the pool and repick.
guidePool <- read.xlsx(paste0(wd, "//preprocessed//200bp_i30.xlsx"))
guidePool <- subset(guidePool, (SNP == "rs8003334|rs8019279|rs8019585")&(sgRNA.Sequence != "AGAACAATAAAAACTACGGT")&(Orientation=="antisense"))
guidePool <- filterGDO(guidePool, minOnScore = 0.2)

# Not a lot of guides, try slightly expanded region.
guidePool <- read.xlsx(paste0(wd, "//preprocessed//400bp_i30.xlsx"))
guidePool <- subset(guidePool, (SNP == "rs8003334|rs8019279|rs8019585")&(sgRNA.Sequence != "AGAACAATAAAAACTACGGT")&(Orientation=="antisense"))
guidePool <- filterGDO(guidePool, minOnScore = 0.2)

# Add this one (antisense): TGGAACTCTCAAGTACTGCC
targetGDO <- bind_rows(targetGDO, subset(guidePool, sgRNA.Sequence=="ACAGGTCTTTGTGGGAACAT"))

```

rs4318227 - adding the guide that downregulated 3 genes in pilot.

```{r}
# Remove the least interesting guide
# ACTGCGCGCAGGGTGGAAGT
removeGDO <- c("ACTGCGCGCAGGGTGGAAGT")
targetGDO <- subset(targetGDO, !(sgRNA.Sequence %in% removeGDO))

# Add in the guide that hit 3 genes
guidePool <- read.xlsx(paste0(wd, "//preprocessed//400bp_i30.xlsx"))
guidePool <- subset(guidePool, (SNP == "rs4318227"))
guidePool <- filterGDO(guidePool, minOnScore = 0.2)

targetGDO <- bind_rows(targetGDO, subset(guidePool, sgRNA.Sequence=="AGAGCTAGACTTGCAAGGGG"))

```

rs1549299 - adding the guide that hit VKORC1 strongly in the 500kb cis analysis
```{r}
# Removing the worst pick order ranked guide
targetGDO <- subset(targetGDO, sgRNA.Sequence != "TCCGGGTGGGGTAGGCACAC")

guidePool <- read.xlsx(paste0(wd, "//preprocessed//400bp_i30.xlsx"))
targetGDO <- bind_rows(targetGDO, subset(guidePool, sgRNA.Sequence=="GCCAAGCGCGGGTTCTGGAA"))
```

rs927174 - adding guide that hit RTF2
```{r}
# GTTCTATGTCTTGTAGCCCA
targetGDO <- subset(targetGDO, sgRNA.Sequence != "GTTCTATGTCTTGTAGCCCA")

guidePool <- read.xlsx(paste0(wd, "//preprocessed//400bp_i30.xlsx"))
targetGDO <- bind_rows(targetGDO, subset(guidePool, sgRNA.Sequence=="AGCATGCCAGGCCTAGATCG"))
```

rs10838702 - adding guide that hit ACP2
```{r}
targetGDO <- subset(targetGDO, sgRNA.Sequence != "TTGGGCTTCAGCTTGAGGAC")

guidePool <- read.xlsx(paste0(wd, "//preprocessed//400bp_i30.xlsx"))
targetGDO <- bind_rows(targetGDO, subset(guidePool, sgRNA.Sequence=="TCTGCAGAATGGCACTGACA"))
```

See adtlRegions for the del region and repicking guides for two regions we are recombining.

```{r}
write.xlsx(targetGDO, file = paste0(wd, "//output//finalTargetGDOi30_wPilot.xlsx"))
```

# Removing guides to problem regions
Will repick using the adtl guides script.
Should remove 12 guides, we will repick 6.
Combine rs4989024 with rs1317708|rs874424
Combine rs2287322 with rs8077638|rs34962442|rs62090051


regions300 definition will combine these guides like this.
```{r}
removeSNP <- c("rs4989024","rs1317708|rs874424", 
            "rs2287322", "rs8077638|rs34962442|rs62090051")

targetGDO <- subset(targetGDO, !(SNP %in% removeSNP) )
```

```{r}
write.xlsx(targetGDO, file = paste0(wd, "//output//finalTargetGDOi30_remBadReg.xlsx"))
```

# Updating Positive Controls
We submitted a list of interesting genes to CRISPick to design promoter targeting control guides.
Next, we want to:
1) Remove any designed guides that overlap with our validated guides.
2) Of the remaining pool, pick the best performing guides. (i.e. top 3 or top 5)

## Load Data
```{r}
wd <- getwd()

# Import new guide designs
posA30 <- read.xlsx(paste0(wd,"//crispick_output//sgrna_designs_promoter_a30.xlsx"))
names(posA30)[names(posA30)=="Target.Gene.Symbol"] <- "SNP"

# Get refseq conversion for chromosome numbers. Needed for compatibility with CRISPick
refseqLookup <- read.delim(paste0(wd,"//input//seq_report_hg19.tsv"))
refseqLookup <- subset(refseqLookup, Role=="assembled-molecule")

# Rename columns for convenience
names(refseqLookup)[names(refseqLookup)=="Chromosome.name"] <- "chr"
names(refseqLookup)[names(refseqLookup)=="RefSeq.seq.accession"] <- "refseq"
refseqLookup <- refseqLookup %>%
  select(chr, refseq)
```

```{r}
# Add chromosome number based on refseq lookup table
posA30 <- addGDOLoc(posA30, refseqLookup)

# Use Input as the "SNP" region
names(posA30)[names(posA30)=="Target.Gene.Symbol"] <- "SNP"

all(posA30$start < posA30$end) # CHECK that coordinate order is correct. TRUE if okay.

# Add rownumber for reference later
posA30$n <- as.numeric(rownames(posA30))


all(posA30$n == as.numeric(rownames(posA30))) #CHECK that n == rownumber. TRUE if okay.
```

```{r}
posGene <- read.table(paste0(wd,"//input//crispick_pos_ctrl.txt"))
colnames(posGene) <- "SNP"

validCtrl <- read.xlsx(paste0(wd, "//input//validated_pos_ctrl.xlsx"))
```

First remove any guides that overlap the validated guides.

```{r}
posOverlap <- findOverlap(validCtrl, posA30,
                          "chr", "start", "end",
                          "chr", "start", "end")
# 12 new guides overlap the validated guides

# Remove guides that overlap validated guides from the pool.
posPool <- posA30[!posA30$n %in% posOverlap[,2],]

```

Pick top 3 guides for each promoter region:
```{r}
# Get indices of guides that are overlapping.
posOverlap <- findSelfOverlap(posPool, "chr", "start", "end")

# Pick top guides for each promoter region.
topPos <- pickTopGDO(3, posA30, posGene, posOverlap)
```

Export the top guides. Go in order of pick order, fill in where we don't have validated guides.
Will combine the lists manually since there's not too many guides.
Need to manually pick+evaluate RPA1/SMYD4 guides since the region is overlapping..
```{r}
write.xlsx(topPos, file = paste0(wd, "//output//top3_promoter_a30.xlsx"))
```

