---
title: "Promoter-Targeting GDO Designs"
author: Shannon Laub
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

if(!require('pacman')) {install.packages('pacman')}
pacman::p_load(openxlsx, dplyr, tidyr, reshape2, stringr)

# BiocManager::install("GenomicRanges")
library(GenomicRanges)

source('libraryDesignFunctions.R')
```

# Updating Positive Controls
We submitted a list of interesting genes to CRISPick to design promoter targeting control guides.
Next, we want to:
1) Remove any designed guides that overlap with our validated guides.
2) Of the remaining pool, pick the best performing guides. (i.e. top 3 or top 5)

## Load Data
```{r}
wd <- getwd()

# Import new guide designs
posA30raw <- read.xlsx(paste0(wd,"//crispick_output//sgrna_designs_promoter_a30.xlsx"))
names(posA30)[names(posA30)=="Target.Gene.Symbol"] <- "SNP"

# Get refseq conversion for chromosome numbers. Needed for compatibility with CRISPick
refseqLookup <- read.delim(paste0(wd,"//input//seq_report_hg19.tsv"))
refseqLookup <- subset(refseqLookup, Role=="assembled-molecule")

# Rename columns for convenience
names(refseqLookup)[names(refseqLookup)=="Chromosome.name"] <- "chr"
names(refseqLookup)[names(refseqLookup)=="RefSeq.seq.accession"] <- "refseq"
refseqLookup <- refseqLookup %>%
  select(chr, refseq)
```

GDO Preprocessing
```{r}
keepCol <- c("Input", "Reference.Sequence","Strand.of.Target", "Orientation", "sgRNA.'Cut'.Position",  
             "sgRNA.Sequence", "sgRNA.Context.Sequence", "On-Target.Efficacy.Score",
             "#.Off-Target.Tier.I.Match.Bin.I.Matches", "#.Off-Target.Tier.II.Match.Bin.I.Matches",
             "#.Off-Target.Tier.III.Match.Bin.I.Matches", "#.Off-Target.Tier.I.Match.Bin.II.Matches",
             "#.Off-Target.Tier.II.Match.Bin.II.Matches", "#.Off-Target.Tier.III.Match.Bin.II.Matches",
             "On-Target.Rank", "Off-Target.Rank",
             "Pick.Order")
posA30 <- select(posA30raw, all_of(keepCol))
posA30 <- filterGDO(posA30)
posA30 <- addGDOLoc(posA30, refseqLookup)
posA30$SNP <- posA30$Input
posA30$n <- as.numeric(rownames(posA30))

# File is now ~same as preprocessed GDOs for target genes. Skipped region assignment since the input indicates the region, here.
write.xlsx(posA30, paste0(wd,"//preprocessed//prom200bp_a30.xlsx"))

# oldPos <- posA30
```

# Pick top GDOs.
```{r}
dirRegion <- paste0(wd,"//input//")
dirGDO <- paste0(wd,"//preprocessed//")

regionList <- c("prom200bp_a30")
fileList <- c("prom200bp_a30")

promGDO <- assembleLibrary(3, fileList, regionList, dirGDO, dirRegion,
                          onTargetFilter = 0.2, pickByStrand = TRUE, overlapLeniency=2)

```
# Substitute for known working guides.
```{r}
# Load dataframe containing working guides from the pilot.
pilot <- read.xlsx(paste0(wd,"//input//validated_pos_ctrl.xlsx"))
pilot <- select(pilot, Target, Sequence, Pick.Order, chr, start, end)
names(pilot) <- c("Input", "sgRNA.Sequence", "Pick.Order", "chr", "start","end")

pilot <- pilot %>%
  mutate(chr = as.character(chr))
pilot$SNP <- pilot$Input
```


```{r}
# the first FERMT2 guide got filtered out for GC content. Let's use it anyways.

fermt2pos <- select(posA30raw, all_of(keepCol))
fermt2pos <- subset(fermt2pos, Input == "FERMT2")
# fermt2pos <- filterGDO(fermt2pos)
fermt2pos <- addGDOLoc(fermt2pos, refseqLookup)
fermt2pos$SNP <- fermt2pos$Input
# what is this?? AGGCCGGCCGGACCCGCTCA #GCTCTGCGGGCGGCGAAGGA
addGuide <- subset(fermt2pos, sgRNA.Sequence == "AGGCCGGCCGGACCCGCTCA")

promGDO <- bind_rows(promGDO, addGuide)
```

```{r}
# Add two of the working TSPAN14 guides from the pilot.
promGDO <- bind_rows(promGDO, subset(pilot, Input == "TSPAN14"))
```

```{r}
# Sub in the working guides for BIN1, CLU, and SNX1
# BIN1 - remove GCCAGGTCCGCGAGGAGGGAZ
# CLU - remove ATCACCACGAATAGCTGTGC
# SNX1 - remove GAGGAGTGGGAGGGCCGCAA
removeGDO <- c("GCCAGGTCCGCGAGGAGGGA", "ATCACCACGAATAGCTGTGC", "GAGGAGTGGGAGGGCCGCAA")
promGDO <- subset(promGDO, !(sgRNA.Sequence %in% removeGDO))

# Add in guides that were validated in the pilot.
addGDO <- c("AAGATCTCCCCGCGCGAGAG", "GGAGCCAGCACAGCTATTCG", "ATTAACGAGTGGGGCCAAAG")
promGDO <- bind_rows(promGDO, subset(pilot, sgRNA.Sequence %in% addGDO))
```

```{r}
promGDO <- promGDO %>%
  arrange(Input, Pick.Order)
```


```{r}
# Pick the top 3 from the RPA1/SMYD4 region
# Also sub in the working guide from the pilot.
keep <- c("RPA1", "SMYD4")
lookat <- subset(promGDO, Input %in% keep)
removeGDO <- c("CCACCGCCAAGACTTCCCCG", "TCCCAAGCGAGGCCTGGCGC", "GCGCTACGCAGCCGCCGCAT")
promGDO <- subset(promGDO, !(sgRNA.Sequence %in% removeGDO))

# Rename the region so it is shared.
promGDO <- promGDO %>%
  mutate(Input = case_when(
    Input == "RPA1" ~ "RPA1|SMYD4",
    Input == "SMYD4" ~ "RPA1|SMYD4",
    TRUE ~ Input
  ))
```

```{r}
length(unique(promGDO$sgRNA.Sequence))
```

# Export pos ctrls
```{r}
write.xlsx(promGDO, paste0(wd,"//output//finalPromoterGDOa30.xlsx"))
```